diff --git a/node_modules/expo-linking/ios/ExpoLinkingModule.swift b/node_modules/expo-linking/ios/ExpoLinkingModule.swift
new file mode 100644
index 0000000..5c958f6
--- /dev/null
+++ b/node_modules/expo-linking/ios/ExpoLinkingModule.swift
@@ -0,0 +1,32 @@
+import ExpoModulesCore
+
+let onURLReceived = "onURLReceived"
+public let onURLReceivedNotification = Notification.Name(onURLReceived)
+
+public class ExpoLinkingModule: Module {
+  public func definition() -> ModuleDefinition {
+    Name("ExpoLinking")
+
+    Events(onURLReceived)
+
+    OnStartObserving(onURLReceived) {
+      NotificationCenter.default.addObserver(self, selector: #selector(handleURLReceivedNotification), name: onURLReceivedNotification, object: nil)
+    }
+
+    OnStopObserving(onURLReceived) {
+      // swiftlint:disable:next notification_center_detachment
+      NotificationCenter.default.removeObserver(self)
+    }
+
+    Function("getLinkingURL") {
+      return ExpoLinkingRegistry.shared.initialURL?.absoluteString
+    }
+  }
+
+  @objc func handleURLReceivedNotification(_ notification: Notification) {
+    guard let url = notification.userInfo?["url"] as? URL else {
+      return
+    }
+    self.sendEvent(onURLReceived, ["url": url.absoluteString])
+  }
+}
